src/sdk/pynni/nni/compression/torch/builtin_quantizers.py

```diff
@@ -29,23 +29,187 @@ class QAT_Quantizer(Quantizer):
+                - op_types : list of string
+                    types of nn.module you want to apply quantization, eg. 'Conv2d'
         """
         super().__init__(model, config_list)
+        self.steps = 0
+        modules_to_compress = self.detect_modules_to_compress()
+        for layer, config in modules_to_compress:
+            if "weight" in config.get("quant_types", []):
+                layer.module.register_buffer("zero_point", None)
+                layer.module.register_buffer("scale", None)
 
-    def quantize_weight(self, weight, config, **kwargs):
-        if config['q_bits'] <= 1:
+    def _range_check(self, tensor):
+        """
+        determine the max value and min value of quantization target.
+
+        Parameters
+        ----------
+        tensor : Tensor
+            the Tensor to be quantized
+
+        Returns
+        -------
+        min : float
+        max : float
+        """
+        # TODO: use EMA to check activation range instead
+        return torch.min(tensor), torch.max(tensor)
+
+    def _update_quantization_param(self, bits, op, rmin, rmax):
+        """
+        update the `zero_point` and `scale` of op.
+
+        Parameters
+        ----------
+        bits : int
+            quantization bits length
+        op : torch.nn.module
+            target module
+        rmin : float
+            min value of real value
+        rmax : float
+            max value of real value
+        """
+        # extend the [min, max] interval to ensure that it contains 0.
+        # Otherwise, we would not meet the requirement that 0 be an exactly
+        # representable value.
+        rmin = min(rmin, 0)
+        rmax = max(rmax, 0)
+
+        # the min and max quantized values, as floating-point values
+        qmin = 0
+        qmax = 1 << bits - 1
+
+        # First determine the scale.
+        scale = (rmax - rmin) / (qmax - qmin)
+
+        # Zero-point computation.
+        initial_zero_point = qmin - rmin / scale
+
+        # Now we need to nudge the zero point to be an integer
+        nudged_zero_point = 0
+        if initial_zero_point < qmin:
+            nudged_zero_point = qmin
+        elif initial_zero_point > qmax:
+            nudged_zero_point = qmax
+        else:
+            nudged_zero_point = torch.round(initial_zero_point)
+
+        op.scale = scale
+        op.zero_point = nudged_zero_point
+
+    def _quantize(self, bits, op, real_val):
+        """
+        quantize real value.
+
+        Parameters
+        ----------
+        bits : int
+            quantization bits length
+        op : torch.nn.module
+            target module
+        real_val : float
+            real value to be quantized
+
+        Returns
+        -------
+        float
+        """
+        transformed_val = op.zero_point + real_val / op.scale
+        qmin = 0
+        qmax = 1 << bits - 1
```

Comment:
Suggest add parentheses here.
This line equals to `1 << (bits - 1)`.

Reply:
fixed